index.js

```diff
@@ -1,83 +1,79 @@
-/* globals Symbol: false, Uint8Array: false, WeakMap: false */
-/*!
- * deep-eql
- * Copyright(c) 2013 Jake Luer <jake@alogicalparadox.com>
- * MIT Licensed
- */
-
-import type from 'type-detect';
-
-function FakeMap() {
-  this._key = 'chai/deep-eql__' + Math.random() + Date.now();
-}
-
-FakeMap.prototype = {
-  get: function get(key) {
-    return key[this._key];
-  },
-  set: function set(key, value) {
-    if (Object.isExtensible(key)) {
-      Object.defineProperty(key, this._key, {
-        value: value,
-        configurable: true,
-      });
-    }
-  },
-};
-
-export var MemoizeMap = typeof WeakMap === 'function' ? WeakMap : FakeMap;
-/*!
- * Check to see if the MemoizeMap has recorded a result of the two operands
+/**
+ * Check to see if the WeakMap has recorded a result of the two operands
  *
  * @param {Mixed} leftHandOperand
  * @param {Mixed} rightHandOperand
- * @param {MemoizeMap} memoizeMap
+ * @param {WeakMap} memoizeMap
  * @returns {Boolean|null} result
-*/
+ */
 function memoizeCompare(leftHandOperand, rightHandOperand, memoizeMap) {
   // Technically, WeakMap keys can *only* be objects, not primitives.
-  if (!memoizeMap || isPrimitive(leftHandOperand) || isPrimitive(rightHandOperand)) {
+  if (
+    !memoizeMap ||
+    isPrimitive(leftHandOperand) ||
+    isPrimitive(rightHandOperand)
+  ) {
     return null;
   }
   var leftHandMap = memoizeMap.get(leftHandOperand);
   if (leftHandMap) {
     var result = leftHandMap.get(rightHandOperand);
-    if (typeof result === 'boolean') {
+    if (typeof result === "boolean") {
       return result;
     }
   }
   return null;
 }
 
+/**
+ * Returns the type of the given variable
+ *
+ * @param {Mixed} obj
+ * @returns {string}
+ */
+function type(obj) {
+  const typeofObj = typeof obj;
+  if (typeofObj !== 'object') {
+    return typeofObj;
+  }
+
+  const toStringLeftSliceLength = 8;
+  const toStringRightSliceLength = -1;
+
+  return Object
```

Comment:
This could return `Symbol.toStringTag` instead, without the slice stuff. I think toStringTag is supported in pretty much everything (except IE which we want to drop support for)

Reply:
It looks like it works for a lot of things but not everything, so I added it but kept the `Object.prototype.toString` for the rest.

Reply:
Here are the tests that failed for me for posterity:

```
  1) Generic
       booleans
         returns false for instances with different values:
     AssertionError: eql(new Boolean(false), new Boolean(true)) === false
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:74:7)
      at process.processImmediate (node:internal/timers:476:21)

  2) Generic
       numbers
         returns false on instances with different signs:
     AssertionError: eql(new Number(-1), new Number(1)) === false
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:155:7)
      at process.processImmediate (node:internal/timers:476:21)

  3) Generic
       dates
         returns false given two dates with the different times:
     AssertionError: eql(dateA, new Date(dateA.getTime() + 1)) === false
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:188:7)
      at process.processImmediate (node:internal/timers:476:21)

  4) Generic
       regexp
         returns false given two regexes with different source:
     AssertionError: eql(/^$/, /^/) === false
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:202:7)
      at process.processImmediate (node:internal/timers:476:21)

  5) Generic
       regexp
         returns false given two regexes with different flags:
     AssertionError: eql(/^/m, /^/i) === false
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:210:7)
      at process.processImmediate (node:internal/timers:476:21)

  6) Generic
       arguments
         returns false given an array:
     AssertionError: eql([], arguments) === false
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:304:7)
      at process.processImmediate (node:internal/timers:476:21)

  7) Generic
       errors
         returns true for errors with same name and message despite different constructors:
     AssertionError: eql(Error("foo"), Object.assign(TypeError("foo"), { name: "Error" }))
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:569:7)
      at process.processImmediate (node:internal/timers:476:21)

  8) Generic
       errors
         returns false for errors with same name but different messages:
     AssertionError: eql(Error("foo"), Error("bar")) === false
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:576:7)
      at process.processImmediate (node:internal/timers:476:21)

  9) Generic
       errors
         returns false for errors with same message but different names:
     AssertionError: eql(Error("foo"), TypeError("foo")) === false
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:583:7)
      at process.processImmediate (node:internal/timers:476:21)

  10) Generic
       errors
         returns true for errors with same name and message despite different otherProp:
     AssertionError: eql(Object.assign(Error("foo"), { otherProp: 42 }), Object.assign(Error("foo"), { otherProp: 13 }))
      at Context.<anonymous> (file:///Users/cr91ew/src/koddsson/deep-eql/test/index.js:626:7)
      at process.processImmediate (node:internal/timers:476:21)
```